pkg/controller/download_request_controller.go

```diff
@@ -0,0 +1,285 @@
+		// we want to shut down the queue via defer and not at the end of the body.
+		wg.Wait()
+
+		glog.Infof("All workers have finished")
+	}()
+
+	glog.Info("Starting DownloadRequestController")
+	defer glog.Infof("Shutting down DownloadRequestController")
+
+	glog.Info("Waiting for caches to sync")
+	if !cache.WaitForCacheSync(ctx.Done(), c.downloadRequestListerSynced, c.backupListerSynced) {
+		return errors.New("timed out waiting for caches to sync")
+	}
+	glog.Info("Caches are synced")
+
+	wg.Add(numWorkers)
+	for i := 0; i < numWorkers; i++ {
+		go func() {
+			wait.Until(c.runWorker, time.Second, ctx.Done())
+			wg.Done()
+		}()
+	}
+
+	wg.Add(1)
+	go func() {
+		wait.Until(c.resync, time.Minute, ctx.Done())
+		wg.Done()
+	}()
+
+	<-ctx.Done()
+
+	return nil
+}
+
+// runWorker runs a worker until the controller's queue indicates it's time to shut down.
+func (c *downloadRequestController) runWorker() {
+	// continually take items off the queue (waits if it's
+	// empty) until we get a shutdown signal from the queue
+	for c.processNextWorkItem() {
+	}
+}
+
+// processNextWorkItem processes a single item from the queue.
+func (c *downloadRequestController) processNextWorkItem() bool {
+	key, quit := c.queue.Get()
+	if quit {
+		return false
+	}
+	// always call done on this item, since if it fails we'll add
+	// it back with rate-limiting below
+	defer c.queue.Done(key)
+
+	err := c.syncHandler(key.(string))
+	if err == nil {
+		// If you had no error, tell the queue to stop tracking history for your key. This will reset
+		// things like failure counts for per-item rate limiting.
+		c.queue.Forget(key)
+		return true
+	}
+
+	glog.Errorf("syncHandler error: %v", err)
+	// we had an error processing the item so add it back
+	// into the queue for re-processing with rate-limiting
+	c.queue.AddRateLimited(key)
+
+	return true
+}
+
+// processDownloadRequest is the default per-item sync handler. It generates a pre-signed URL for
+// a new DownloadRequest or deletes the DownloadRequest if it has expired.
+func (c *downloadRequestController) processDownloadRequest(key string) error {
+	glog.V(4).Infof("processDownloadRequest for key %q", key)
+	ns, name, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		glog.V(4).Infof("error splitting key %q: %v", key, err)
+		return err
+	}
+
+	downloadRequest, err := c.downloadRequestLister.DownloadRequests(ns).Get(name)
+	if apierrors.IsNotFound(err) {
+		glog.V(4).Infof("unable to find DownloadRequest %q", key)
+		return nil
+	}
+	if err != nil {
+		glog.V(4).Infof("error getting DownloadRequest %q: %v", key, err)
+		return err
+	}
+
+	switch downloadRequest.Status.Phase {
+	case "", v1.DownloadRequestPhaseNew:
+		return c.generatePreSignedURL(downloadRequest)
+	case v1.DownloadRequestPhaseProcessed:
+		return c.deleteIfExpired(downloadRequest)
+	}
+
+	return nil
+}
+
+const signedURLTTL = 10 * time.Minute
```

Comment:
One general question I have, which I can test if you aren't sure is: If the TTL expires on the actual signed URL in the middle of an in process download (slow connection, large download, etc.) will it proceed or be terminated?

Reply:
I don't know